include/dkm_matrix.hpp

```diff
@@ -0,0 +1,63 @@
+#pragma once
+
+#ifndef DKM_MATRIX_HPP
+#define DKM_MATRIX_HPP
+
+#include <vector>
+#include <cassert>
+#include <algorithm>
+
+namespace dkm
+{
+// This class is only an interface! Not designed to be used outside library internals.
+
+template<typename T>
+class as_matrix
+{
+    const T *data = nullptr;
+
+    // column major indexer
+    auto cm_indexer(size_t i, size_t j) const -> const T&
+    {
+        assert(i >= 0 && j >= 0 && i < n_rows && j < n_cols);
+        return data[j * n_rows + i];
+    }
+
+     // row major indexer
+    auto rm_indexer(size_t i, size_t j) const -> const T&
+    {
+        assert(i >= 0 && j >= 0 && i < n_rows && j < n_cols);
+        return data[i * n_cols + j];
+    }
+
+    const T& (as_matrix<T>::*indexer) (size_t, size_t) const = nullptr;
+
+public:
+    const size_t n_rows, n_cols;
+
+    as_matrix(const T *data, size_t n_rows, size_t n_cols, bool col_major = true)
+        : data(data), n_rows(n_rows), n_cols(n_cols),
+          indexer((col_major)? &as_matrix<T>::cm_indexer: &as_matrix<T>::rm_indexer)
+    {}
+
+    auto row(size_t i) const -> std::vector<T>;
+    auto operator()(size_t i, size_t j) const -> const T&
+    {
+        return ((this)->*(this->indexer))(i, j);
+    }
+};
+
+
+template<typename T>
+auto as_matrix<T>::row(size_t i) const -> std::vector<T>
+{
+    auto res = std::vector<T>(n_cols);
+    for(size_t j = 0; j < n_cols; j++)
+    {
+        res[j] = (*this)(i, j);
```

Comment:
One of the reasons this library is so performant is that it avoids allocations and copies wherever possible. This is a full copy of each and every row in the input data, multiple times. It _will_ have a significant effect on performance.

The way to do this performantly and safely would be to return a [`std::span`](https://en.cppreference.com/w/cpp/container/span) which points to the internal data. Given this library currently only requires C++11, the solution is probably returning a pointer to the row or a custom struct which emulates the behavior of `std::span`.

Reply:
I had been implementing a new version that takes this into account. The idea is whenever the matrix is column major we first  copy (because we cannot change the incoming data) and transpose it.

The copy would work like this:
```C++
auto [owner, data] = copy(matrix);
```
From here we can safely return the span. See that the matrix data structure never owns the data, so we need to return an owner (arguably `unique_ptr`) from the copy function.

I plan to implement this when my project reaches the optimization stage.  Here is an implementation I was already working on https://pastebin.com/DYQs2EBb

Reply:
Why can transposition not just produce a new `as_matrix` with the correct indexer if we need to switch to column instead of row major.

There's no reason to have any solution here that includes any form of copying.

Reply:
My argument was mainly that in this case the representation of the data doesn't need to change at all, we just need need a different view over the data, which seems to be the whole point of `as_matrix` (an abstracted view over the data).

Reply:
I understand your point, my only concern is that in the case of row major data to get a row we can easily return `std::span(ptr, ncols)`. But in case of column major it is not possible without adding more complexity. To avoid copying the alternative I can think of would have to be a sort of `row_vec` whose the indexing would work as following:
```C++
auto row_vec::operator()(size_t i)
{
    return data[i * n_rows + this->row_number]
}
```
From my experience the transposing the data + changing from  column major to row major is simpler and faster, specially when the input data is large enough. But I am not completely sure if this is true in fact. What is your opinion on this?